kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala

```diff
@@ -606,6 +606,36 @@ object KyuubiConf {
     .version("1.2.0")
     .fallbackConf(LEGACY_ENGINE_SHARE_LEVEL)
 
+  val ENGINE_PROVIDE_POLICY: ConfigEntry[String] = buildConf("engine.provide.policy")
+    .doc("Multiple engines can be exposed under same space when using engine pool, " +
+      "choose the appropriate strategy according to different scenarios.")
+    .version("1.4.0")
+    .stringConf
+    .transform(_.toUpperCase(Locale.ROOT))
+    .createWithDefault(ProvidePolicy.RANDOM.toString)
+
+  val ENGINE_POOL_ENABLED: ConfigEntry[Boolean] = buildConf("engine.pool.enabled")
+    .doc("When true, kyuubi will provides engine pool feature.")
+    .version("1.4.0")
+    .booleanConf
+    .createWithDefault(false)
+
+  val ENGINE_POOL_SIZE_THRESHOLD: ConfigEntry[Int] = buildConf("engine.pool.size.threshold")
+    .doc("This parameter is introduced as a server-side parameter, " +
+      "and controls the upper limit of the engine pool.")
+    .version("1.4.0")
+    .intConf
+    .checkValue(s => s > 0 && s <= 32, "Invalid Engine Pool Threshold Size, " +
+      "should be between 0 and 33.")
+    .createWithDefault(16)
+
+  val ENGINE_POOL_SIZE: ConfigEntry[Int] = buildConf("engine.pool.size")
+    .doc("Max number of engines that can be provided to clients.")
+    .version("1.4.0")
+    .intConf
+    .checkValue(s => s > 0, "Invalid Engine Pool Size, must be positive number.")
+    .createWithDefault(1)
```

Comment:
I have a question about why split it into two parameters here?

Reply:
I think the option `kyuubi.engine.pool.size.threshold` cat give users the opportunity to set the threshold.

if we use `checkValue`, the `50` is hardcode.
`.checkValue(s => s > 0 && s < 50, "Invalid Engine Pool Size, should be between 0 and 50")`

Reply:
I can accept this, but why the threshold of ENGINE_POOL_SIZE_THRESHOLD is 32?

Reply:
For the threshold`32`, willing to talk with community.

Reply:
How about?

for server-side config `engine.pool.size.threshold`, we can leave it to the system administrators to set, I vote positive numbers as valid values and maybe 9(which also means `kyuu`LOL) as default

 for user-faced config `engine.pool.size` maybe we need not check it, and use -1 by default to turn off the pool featuer

Reply:
agree.